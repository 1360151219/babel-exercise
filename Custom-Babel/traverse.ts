const astDefinationsMap = new Map();
astDefinationsMap.set('Program', {
    visitor: ['body']
});
astDefinationsMap.set('VariableDeclaration', {
    visitor: ['declarations']
});
astDefinationsMap.set('VariableDeclarator', {
    visitor: ['id', 'init']
});
astDefinationsMap.set('Identifier', {});
astDefinationsMap.set('NumericLiteral', {});
astDefinationsMap.set('FunctionDeclaration', {
    visitor: ['id', 'params', 'body']
});
astDefinationsMap.set('BlockStatement', {
    visitor: ['body']
});
astDefinationsMap.set('ReturnStatement', {
    visitor: ['argument']
});
astDefinationsMap.set('BinaryExpression', {
    visitor: ['left', 'right']
});
astDefinationsMap.set('ExpressionStatement', {
    visitor: ['expression']
});
astDefinationsMap.set('CallExpression', {
    visitor: ['callee', 'arguments']
});

export default function traverse(node, visitor) {
    const curNodeType = node.type
    if(!astDefinationsMap.has(curNodeType)){
        throw new Error("error type");
    }
    visitor[curNodeType] && visitor[curNodeType](node)
    const nextVisitors = astDefinationsMap.get(curNodeType)?.visitor;
    if(!nextVisitors){
        // 不需要再遍历下去了
        return
    }    
    nextVisitors.forEach(props => {
        const nextNode = node[props]
        if (Array.isArray(nextNode)) {
            nextNode.forEach(n => {
                traverse(n, visitor)
            })
        } else {
            traverse(nextNode, visitor)
        }
    });
}