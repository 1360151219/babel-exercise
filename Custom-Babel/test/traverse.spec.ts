import { expect, it } from 'vitest'
import Parser from '../index'
import traverse from '../traverse'
function parse(code: string) {
  return Parser.parse(code, {
    locations: true,
    ecmaVersion: 'latest'
  })
}
it("traverse",()=>{
  const code = 'let a = 1';
  const ast = parse(code)
  traverse(ast,{
    Identifier:(node)=>{
      node.name = 'b'
    }
  })
  expect(ast).toMatchSnapshot("traverse")
})