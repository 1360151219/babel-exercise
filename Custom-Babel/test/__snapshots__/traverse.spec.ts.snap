// Vitest Snapshot v1

exports[`traverse > traverse 1`] = `
Node {
  "body": [
    Node {
      "declarations": [
        Node {
          "end": 9,
          "id": Node {
            "end": 5,
            "loc": SourceLocation {
              "end": Position {
                "column": 5,
                "line": 1,
              },
              "start": Position {
                "column": 4,
                "line": 1,
              },
            },
            "name": "b",
            "start": 4,
            "type": "Identifier",
          },
          "init": Node {
            "end": 9,
            "loc": SourceLocation {
              "end": Position {
                "column": 9,
                "line": 1,
              },
              "start": Position {
                "column": 8,
                "line": 1,
              },
            },
            "raw": "1",
            "start": 8,
            "type": "NumericLiteral",
            "value": 1,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 9,
              "line": 1,
            },
            "start": Position {
              "column": 4,
              "line": 1,
            },
          },
          "start": 4,
          "type": "VariableDeclarator",
        },
      ],
      "end": 9,
      "kind": "let",
      "loc": SourceLocation {
        "end": Position {
          "column": 9,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "start": 0,
      "type": "VariableDeclaration",
    },
  ],
  "end": 9,
  "loc": SourceLocation {
    "end": Position {
      "column": 9,
      "line": 1,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;
